File: mockserver-client/src/main/java/org/mockserver/client/MockServerClient.java
package org.mockserver.client;

import com.google.common.annotations.VisibleForTesting;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.client.util.StringContentProvider;
import org.eclipse.jetty.http.HttpMethod;
import org.mockserver.client.serialization.ExpectationSerializer;
import org.mockserver.matchers.Times;
import org.mockserver.mock.Expectation;
import org.mockserver.model.HttpRequest;
import org.mockserver.model.HttpResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author jamesdbloom
 */
public class MockServerClient {
    private final Logger logger = LoggerFactory.getLogger(this.getClass());

    private final String mockServerURI;
    private HttpClient httpClient = new HttpClient();

    private ExpectationSerializer expectationSerializer = new ExpectationSerializer();

    /**
     * Start the client communicating to a MockServer at the specified host and port
     * for example:
     * <p/>
     * MockServerClient mockServerClient = new MockServerClient("localhost", 8080);
     *
     * @param host the host for the MockServer to communicate with
     * @param port the port for the MockServer to communicate with
     */
    public MockServerClient(String host, int port) {
        mockServerURI = "http://" + host + ":" + port + "/";
        try {
            httpClient.start();
        } catch (Exception e) {
            logger.error("Exception starting HttpClient in MockServerClient", e);
            throw new RuntimeException("Exception starting HttpClient in MockServerClient", e);
        }
    }

    @VisibleForTesting
    MockServerClient(String host, int port, HttpClient httpClient) {
        mockServerURI = "http://" + host + ":" + port + "/";
        this.httpClient = httpClient;
        try {
            this.httpClient.start();
        } catch (Exception e) {
            logger.error("Exception starting HttpClient in MockServerClient", e);
            throw new RuntimeException("Exception starting HttpClient in MockServerClient", e);
        }
    }


    /**
     * Specify an unlimited expectation that will respond regardless of the number of matching request
     * for example:
     * <p/>
     * mockServerClient
     * .when(
     * new HttpRequest()
     * .withPath("/some_path")
     * .withBody("some_request_body")
     * )
     * .respond(
     * new HttpResponse()
     * .withBody("some_response_body")
     * .withHeaders(
     * new Header("responseName", "responseValue")
     * )
     * );
     *
     * @param httpRequest the request that must be matched for this expectation to respond
     * @return an Expectation object that can be used to specify the response
     */
    public ForwardChainExpectation when(final HttpRequest httpRequest) {
        return when(httpRequest, Times.unlimited());
    }

    /**
     * Specify an limited expectation that will respond a specified number of times when the request is matched
     * for example:
     * <p/>
     * mockServerClient
     * .when(
     * new HttpRequest()
     * .withPath("/some_path")
     * .withBody("some_request_body"),
     * Times.exactly(5)
     * )
     * .respond(
     * new HttpResponse()
     * .withBody("some_response_body")
     * .withHeaders(
     * new Header("responseName", "responseValue")
     * )
     * );
     *
     * @param httpRequest the request that must be matched for this expectation to respond
     * @param times       the number of times to respond when this request is matched
     * @return an Expectation object that can be used to specify the response
     */
    public ForwardChainExpectation when(HttpRequest httpRequest, Times times) {
        return new ForwardChainExpectation(this, new Expectation(httpRequest, times));
    }

    /**
     * Pretty-print the json for all expectations already setup to the log.  They are printed at
     * WARN level to ensure they appear even if the default logging level has not been altered
     */
    public void dumpToLog() {
        try {
            sendRequest("", "dumpToLog");
        } catch (Exception e) {
            logger.error("Exception sending reset request to MockServer", e);
            throw new RuntimeException("Exception sending reset request to MockServer", e);
        }
    }

    /**
     * Reset MockServer by clearing all expectations
     */
    public void reset() {
        try {
            sendRequest("", "reset");
        } catch (Exception e) {
            logger.error("Exception sending reset request to MockServer", e);
            throw new RuntimeException("Exception sending reset request to MockServer", e);
        }
    }

    /**
     * Clear all expectations that match the request
     *
     * @param httpRequest the request that is matched against when deciding whether to clear each expectation
     */
    public void clear(HttpRequest httpRequest) {
        sendExpectation(new Expectation(httpRequest, Times.unlimited()).respond(new HttpResponse()), "clear");
    }

    protected void sendExpectation(Expectation expectation) {
        sendExpectation(expectation, "");
    }

    private void sendExpectation(Expectation expectation, String path) {
        try {
            sendRequest(expectation != null ? expectationSerializer.serialize(expectation) : "", path);
        } catch (Exception e) {
            logger.error(String.format("Exception sending expectation to MockServer as %s", expectation), e);
            throw new RuntimeException(String.format("Exception sending expectation to MockServer as %s", expectation), e);
        }
    }

    private void sendRequest(String body, String path) throws Exception {
        httpClient.newRequest(mockServerURI + path)
                .method(HttpMethod.PUT)
                .header("Content-Type", "application/json; charset=utf-8")
                .content(new ComparableStringContentProvider(body, "UTF-8"))
                .send();
    }

    static class ComparableStringContentProvider extends StringContentProvider {

        public ComparableStringContentProvider(String content, String encoding) {
            super(content, encoding);
        }

        @Override
        public boolean equals(Object other) {
            return EqualsBuilder.reflectionEquals(this, other);
        }

        @Override
        public int hashCode() {
            return HashCodeBuilder.reflectionHashCode(this);
        }
    }
}

File: mockserver-client-javascript/src/main/javascript/mockServerClient.js
/**
 * Start the client communicating to a MockServer at the specified host and port
 * for example:
 *
 *   var client = mockServerClient("localhost", 1080);
 *
 * @param host the host for the MockServer to communicate with
 * @param port the port for the MockServer to communicate with
 */
var mockServerClient = function (host, port) {
    "use strict";

    var xmlhttp = new XMLHttpRequest();
    var mockServerUrl = "http://" + host + ":" + port;
    /**
     * The default headers added to to the mocked response when using mockSimpleResponse(...)
     */
    var defaultResponseHeaders = [
        {"name": "Content-Type", "values": ["application/json; charset=utf-8"]},
        {"name": "Cache-Control", "values": ["no-cache, no-store"]}
    ];
    var createResponseMatcher = function (path) {
        var headers = [];
        if (window.location.href.match(/testId\=.*/g)) {
            headers = [
                {
                    "name": "Referer",
                    "values": [".*" + window.location.href.match(/testId\=.*/g)[0] + ".*"]
                }
            ];
        }
        return {
            method: "",
            path: path,
            queryStringParameters: [],
            body: "",
            headers: headers,
            cookies: []
        }
    };
    var createExpectation = function (path, responseBody, statusCode) {
        return {
            httpRequest: createResponseMatcher(path),
            httpResponse: {
                statusCode: statusCode || 200,
                body: JSON.stringify(responseBody),
                cookies: [],
                headers: defaultResponseHeaders,
                delay: {
                    timeUnit: "MICROSECONDS",
                    value: 0
                }
            },
            times: {
                remainingTimes: 1,
                unlimited: false
            }
        };
    };
    /**
     * Setup an expectation in the MockServer by specifying an expectation object
     * for example:
     *
     *   mockServerClient("localhost", 1080).mockAnyResponse(
     *       {
     *           'httpRequest': {
     *               'path': '/somePath',
     *               'body': {
     *                   'type': "STRING",
     *                   'value': 'someBody'
     *               }
     *           },
     *           'httpResponse': {
     *               'statusCode': 200,
     *               'body': Base64.encode(JSON.stringify({ name: 'first_body' })),
     *               'delay': {
     *                   'timeUnit': 'MILLISECONDS',
     *                   'value': 250
     *               }
     *           },
     *           'times': {
     *               'remainingTimes': 1,
     *               'unlimited': false
     *           }
     *       }
     *   );
     *
     * @param expectation the expectation to setup on the MockServer
     */
    var mockAnyResponse = function (expectation) {
        xmlhttp.open("PUT", mockServerUrl + "/expectation", false);
        xmlhttp.setRequestHeader("Content-Type", "application/json; charset=utf-8");
        xmlhttp.send(JSON.stringify(expectation));
        return _this;
    };
    /**
     * Setup an expectation in the MockServer without having to specify the full expectation object
     * for example:
     *
     *   mockServerClient("localhost", 1080).mockSimpleResponse('/somePath', { name: 'value' }, 203);
     *
     * @param path the path to match requests against
     * @param responseBody the response body to return if a request matches
     * @param statusCode the response code to return if a request matches
     */
    var mockSimpleResponse = function (path, responseBody, statusCode) {
        return mockAnyResponse(createExpectation(path, responseBody, statusCode));
    };
    /**
     * Override the default headers that are used to specify the response headers in mockSimpleResponse(...)
     * (note: if you use mockAnyResponse(...) the default headers are not used)
     * for example:
     *
     *   mockServerClient("localhost", 1080).setDefaultHeaders([
     *       {"name": "Content-Type", "values": ["application/json; charset=utf-8"]},
     *       {"name": "Cache-Control", "values": ["no-cache, no-store"]}
     *   ])
     *
     * @param headers the path to match requests against
     */
    var setDefaultHeaders = function (headers) {
        defaultResponseHeaders = headers;
    };
    /**
     * Verify a request has been sent for example:
     *
     *   client.verify({
     *      'method': 'POST',
     *      'path': '/somePath'
     *   });
     *
     * @param request the http request that must be matched for this verification to pass
     * @param count   the number of times this request must be matched
     * @param exact   true if the count is matched as "equal to" or false if the count is matched as "greater than or equal to"
     */
    var verify = function (request, count, exact) {
        if (count === undefined) {
            count = 1;
        }
        xmlhttp.open("PUT", mockServerUrl + "/verify", false);
        xmlhttp.setRequestHeader("Content-Type", "application/json; charset=utf-8");
        xmlhttp.send(JSON.stringify({
            "httpRequest": request,
            "times": {
                "count": count,
                "exact": exact
            }
        }));
        if (xmlhttp.status !== 202) {
            console && console.error && console.error(xmlhttp.responseText);
            throw xmlhttp.responseText;
        }
        return _this;
    };
    /**
     * Verify a sequence of requests has been sent for example:
     *
     *   client.verifySequence(
     *       {
     *          'method': 'POST',
     *          'path': '/first_request'
     *       },
     *       {
     *          'method': 'POST',
     *          'path': '/second_request'
     *       },
     *       {
     *          'method': 'POST',
     *          'path': '/third_request'
     *       }
     *   );
     *
     * @param arguments the list of http requests that must be matched for this verification to pass
     */
    var verifySequence = function () {
        xmlhttp.open("PUT", mockServerUrl + "/verifySequence", false);
        xmlhttp.setRequestHeader("Content-Type", "application/json; charset=utf-8");
        var requestSequence = [];
        for (var i = 0; i < arguments.length; i++) {
            requestSequence.push(arguments[i]);
        }
        xmlhttp.send(JSON.stringify({
            "httpRequests": requestSequence
        }));
        if (xmlhttp.status !== 202) {
            console && console.error && console.error(xmlhttp.responseText);
            throw xmlhttp.responseText;
        }
        return _this;
    };
    /**
     * Reset MockServer by clearing all expectations
     */
    var reset = function () {
        xmlhttp.open("PUT", mockServerUrl + "/reset", false);
        xmlhttp.send("");
        return _this;
    };
    /**
     * Clear all expectations that match the specified path
     *
     * @param pathOrRequestMatcher  if a string is passed in the value will be treated as the path to
     *                              decide which expectations to cleared, however if an object is passed
     *                              in the value will be treated as a full request matcher object
     */
    var clear = function (pathOrRequestMatcher) {
        xmlhttp.open("PUT", mockServerUrl + "/clear", false);
        xmlhttp.setRequestHeader("Content-Type", "application/json; charset=utf-8");
        if (typeof pathOrRequestMatcher === "string") {
            xmlhttp.send(JSON.stringify(createResponseMatcher(pathOrRequestMatcher)));
        } else if (pathOrRequestMatcher) {
            xmlhttp.send(JSON.stringify(pathOrRequestMatcher));
        } else {
            xmlhttp.send(JSON.stringify(createResponseMatcher(".*")));
        }
        return _this;
    };
    /**
     * Retrieve the recorded requests that match the parameter, as follows:
     * - use a string value to match on path,
     * - use a request matcher object to match on a full request,
     * - or use null to retrieve all requests
     *
     * @param pathOrRequestMatcher  if a string is passed in the value will be treated as the path, however
     *                              if an object is passed in the value will be treated as a full request
     *                              matcher object, if null is passed in it will be treated as match all
     */
    var retrieveRequests = function (pathOrRequestMatcher) {
        xmlhttp.open("PUT", mockServerUrl + "/retrieve", false);
        xmlhttp.setRequestHeader("Content-Type", "application/json; charset=utf-8");
        if (typeof pathOrRequestMatcher === "string") {
            xmlhttp.send(JSON.stringify(createResponseMatcher(pathOrRequestMatcher)));
        } else if (pathOrRequestMatcher) {
            xmlhttp.send(JSON.stringify(pathOrRequestMatcher));
        } else {
            xmlhttp.send(JSON.stringify(createResponseMatcher(".*")));
        }
        return JSON.parse(xmlhttp.responseText);
    };
    /**
     * Retrieve the setup expectations that match the parameter,
     * the expectation is retrieved by matching the parameter
     * on the expectations own request matcher, as follows:
     * - use a string value to match on path,
     * - use a request matcher object to match on a full request,
     * - or use null to retrieve all requests
     *
     * @param pathOrRequestMatcher  if a string is passed in the value will be treated as the path, however
     *                              if an object is passed in the value will be treated as a full request
     *                              matcher object, if null is passed in it will be treated as match all
     */
    var retrieveExpectations = function (pathOrRequestMatcher) {
        xmlhttp.open("PUT", mockServerUrl + "/retrieve?type=expectation", false);
        xmlhttp.setRequestHeader("Content-Type", "application/json; charset=utf-8");
        if (typeof pathOrRequestMatcher === "string") {
            xmlhttp.send(JSON.stringify(createResponseMatcher(pathOrRequestMatcher)));
        } else if (pathOrRequestMatcher) {
            xmlhttp.send(JSON.stringify(pathOrRequestMatcher));
        } else {
            xmlhttp.send(JSON.stringify(createResponseMatcher(".*")));
        }
        return JSON.parse(xmlhttp.responseText);
    };

    var _this = {
        mockAnyResponse: mockAnyResponse,
        mockSimpleResponse: mockSimpleResponse,
        setDefaultHeaders: setDefaultHeaders,
        verify: verify,
        verifySequence: verifySequence,
        reset: reset,
        clear: clear,
        retrieveRequests: retrieveRequests,
        retrieveExpectations: retrieveExpectations
    };
    return _this;
};